### PR TITLE
fix(field validation): don't validate disabled components; guard calls to validate and resetValidation #8481

### PR DESCRIPTION
**Pull request by pdanpdan**
#8481

### CODE REVIEW

File: ui/src/components/form/QForm.js [hunk 1]
```diff
@@ -49,7 +49,7 @@ export default Vue.extend({
 
       for (let i = 0; i < components.length; i++) {
         const comp = components[i]
-        const valid = comp.validate()
+        const valid = typeof comp.validate === 'function' ? comp.validate() : true
```
Comment (rstoenescu) [inline ui/src/components/form/QForm.js:52]: comp is ensured to have validate() from this.getValidationComponents()

File: ui/src/components/form/QForm.js [hunk 2]
```diff
@@ -113,7 +113,7 @@ export default Vue.extend({
       this.validateIndex++
 
       this.getValidationComponents().forEach(comp => {
-        comp.resetValidation()
+        typeof comp.resetValidation === 'function' && comp.resetValidation()
```
Comment (rstoenescu) [inline ui/src/components/form/QForm.js:116]: not having resetValidation() would be a user-error. This is required.

File: ui/src/mixins/validate.js [hunk 3]
```diff
@@ -34,6 +34,15 @@ export default {
...
 
+    disable (val) {
+      if (val === true) {
+        this.resetValidation()
+      }
+      else {
+        this.__validateIfNeeded(true)
+      }
+    },
+
```
Comment (rstoenescu) [inline ui/src/mixins/validate.js:45]: When passing from "enable" to "disable" it makes sense to resetValidation().
But not the other way around. Let the submit trigger this, or whatever the developer has configured. So please remove the `this.__validateIfNeeded(true)`

**Review by rstoenescu — CHANGES_REQUESTED** [review]

File: ui/src/components/form/QForm.js (hunk 1 above)
Comment (pdanpdan) [reply ui/src/components/form/QForm.js:52]: My bad, it's checked.

File: ui/src/components/form/QForm.js (hunk 2 above)
Comment (pdanpdan) [reply ui/src/components/form/QForm.js:116]: There is no documentation I can find about it being required. Even we have in dev at least one case like this.
And I can think of some reasons not to implement it :)
The problem is that if it's not present it all crashes and it's not easy to see why.

File: ui/src/mixins/validate.js (hunk 3 above)
Comment (pdanpdan) [reply ui/src/mixins/validate.js:45]: ok

File: ui/src/components/form/QForm.js (hunk 2 above)
Comment (pdanpdan) [reply ui/src/components/form/QForm.js:116]: As this is something seldom called , I think the benefits of this being checked are worth it.

**Comment by rstoenescu** [comment]
Please check vue3-work for porting the same idea. There's two methods that I've designed there (a mixin for Options API and a composable for Composition API). Both work differently than Quasar v1. @pdanpdan

**Pull request merged**